server/src/main/java/org/elasticsearch/plugins/PluginsLoader.java

```diff
@@ -161,6 +160,13 @@ public PluginsLoader(Path modulesDirectory, Path pluginsDirectory) {
         this.loadedPluginLayers = Collections.unmodifiableMap(loadPluginLayers(seenBundles, qualifiedExports));
     }
 
+    // Used only for testing.
+    PluginsLoader() {
```

Comment:
This is only here because of `TestPluginsLoader()`, correct?
I'd prefer to have an internal `PluginsLoader` ctor that just sets the finals (it just does three assignements,  `this.moduleDescriptors = moduleDescriptors` etc.) and both TestPluginsLoader()` and `PluginsLoader(Path modulesDirectory, Path pluginsDirectory)` calling it.
The first will simply pass `Collections.emptyList()`, the second will use some static helpers to build the collections (like we already do with `loadPluginLayers()` for `this.loadedPluginLayers`).

We can even go a step further and eliminate the call to `addServerExportsService` by passing down the `qualifiedExports` map to another "internal" ctor, passing down the result of a method similart to  `addServerExportsService` (but static) from the current `PluginsLoader(Path modulesDirectory, Path pluginsDirectory)`  ctor.
`TestPluginsLoader` instead will pass down an an empty map. That would make having `TestPluginsLoader` worthwhile, I think.

Reply:
@ldematte I've updated my PR to match what I think you've written here. Please let me know if I misinterpreted anything. Turns out with the changes `TestPluginsLoader` is no longer necessary since we can just use the second constructor you've suggested here.